Move ToServices translation into operator

tgraf opened the pull request:
The translation of ToServices rules to the corresponding service backend CIDRs can be done in the operator similar to how ToGroups are being resolved.

Remove all of the rule translation code from the agent itself and avoid triggering a full policy recalculation on every service update when ToServices rules are in effect.

This removes a quite horrible `TriggerPolicyUpdate()` call and hides everything behind standard CIDR updates which can be optimized locally with local identities.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6509)
<!-- Reviewable:end -->

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -878,6 +896,24 @@ func (kub *Kubectl) ciliumInstall(dsPatchName, cmPatchName string, getK8sDescrip
+		}
+
+		ginkgoext.By("Waiting for operator to become ready")
+		err := kub.WaitforNPods(KubeSystemNamespace, "-l name=cilium-operator", 1, 300)
```
I would not wait here, for wait I would add that into `ExpectCiliumReady`  due is master all needs the operator. 

Also, I would create `ExpectCiliumReadyVersion(version)` to validate in the upgrade test here: 
https://github.com/cilium/cilium/blob/master/test/k8sT/Updates.go#L152

The reason, in some cases we install Cilium + ETCD operator + dns and we install all, and wait after that. In this case can produce a chicken egg issue. 

Just a comment, feel free to change or not.

tgraf replied:
What is the reason for the split? I understand that everything has to be deployed before we start waiting for things to come up but why does the readiness check have to be called separately. What if the user forgets?

eloycoto replied:
I understand that etcd-operator/kube-dns/cilium is a separated thing, with the expect after that you can fail with more context. 

> What if the user forgets?

I'm not sure If I get this one

tgraf replied:
My main point is that it is quite hard to understand that `ExpectCiliumReady()` has to be called afterwards but only after `DeployETCDOperator()` has been called first. I don't fully understand why these functions are split apart and why we don't just have a single install function that ensures proper installation and fails otherwise.

Anyway, I've removed the check as it is mostly cosmetic. The operator can come up in the back while the tests are already running.

eloycoto reviewed (dismissed):
Few concerns shared offline. 
Minor comment related with WaitForPods, just a style comment, feel free to change. 

The only thing that it's required is to add operator on `test/kubernetes_test.sh`

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -845,6 +855,14 @@ func (kub *Kubectl) ciliumInstall(dsPatchName, cmPatchName string, getK8sDescrip
+	operatorPathname := getK8sDescriptor("cilium-operator.yaml")
+	if versionRequiresOperator(versionTag) {
+		if operatorRbacPathname == "" || operatorPathname == "" {
+			return fmt.Errorf("Cilium DaemonSet descriptor not found")
```
Nit: should say "Cilium operator descriptor not found"

ianvernon reviewed (approved):
This is an awesome change!

tgraf committed:
kubernetes: Fix operator ClusterRole and ClusterRoleBinding ApiVersion

Use the template keywords to translate into the correct apiVersion depending on
the Kubenernetes version

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
kubernetes: Split operator YAML file to allow patching

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
operator: Move ToServices translations into operator

The translation of ToServices rules to the corresponding service backend CIDRs
can be done in the operator similar to how ToGroups are being resolved.

Remove all of the rule translation code from the agent itself and avoid
triggering a full policy recalculation on every service update when ToServices
rules are in effect.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
operator: Move CNP specific handling into operator

The pkg/policy/groups package was polluted by Kubernetes specific CNP handling.
Move all of this into operator/ so it can be reused for services translation.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
test: Build & run cilium-operator

Build a cilium-operator image based on the local source tree and push it to the
local image registry.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
operator: Listen to services changes even if service synchronization is disabled

This allows to use the ServiceCache for CNP translation.

Signed-off-by: Thomas Graf <thomas@cilium.io>
